runtimevar/runtimevar.go

```diff
@@ -124,80 +128,190 @@ var (
 	}
 )
 
+type snapshotPlusError struct {
+	snapshot Snapshot
+	err      error
+}
+
 // Variable provides an easy and portable way to watch runtime configuration
 // variables. To create a Variable, use constructors found in provider-specific
 // subpackages.
 type Variable struct {
-	watcher  driver.Watcher
+	dw       driver.Watcher
 	provider string // for metric collection
-	nextCall time.Time
-	prev     driver.State
+
+	// For cancelling the background goroutine, and noticing when it has exited.
+	backgroundCancel  func()
+	backgroundCloseCh chan bool
+
+	// haveGoodCh is closed and haveGood is set to true when we get the first
+	// good value for the variable.
+	haveGoodCh chan bool
+	haveGood   bool
+
+	// Holds the snapshot/error to return from the next call to Watch.
+	// The buffer size is 1 so that we can store something in it even when there's
+	// not a Watch in flight.
+	nextWatchCh chan *snapshotPlusError
```

Comment:
> the next read can be really out of date if Watch isn't called often enough

Why is that a problem? `Watch` always returns a change when one happens, and always returns the latest change (because of the code that does a non-blocking read from the channel; see line 265) . It's possible that if you don't call `Watch` enough, you will miss some changes, but I'm not sure why that's a problem (especially given that driver implementations will often miss changes too, e.g. the polling ones).

Reply:
It can be made to work, but my instinct is that trying to get this approach to be correct will be more costly than the alternative, which also has better behavior about getting stale reads. Using an approach like the above has less parts to communicate the same information.

For instance, below, `Watch` does not get interrupted if there is a `Close` during a blocking `Watch`. It only checks at the beginning.

Reply:
It does get interrupted, because `Close` closes the `nextWatchCh`. It wasn't checking for that case though, so it panicked. I fixed that and added a test.

Reply:
Let's chat in person about this tomorrow. I am still not sure what you mean by "stale reads", and I haven't been able to figure out how it works:

* Won't the background goroutine block when writing to the `next` channel if nobody is calling `Latest` or `Watch`? Unless it's buffered, but then how big should the buffer be, and doesn't that cause `Watch`/`Latest` to return old data?
* I don't understand why `Latest` would block on `next` every time. Once there's a good value, we want to return it right away.

Reply:
Discussed offline. Agreed that using a `sync.Mutex` is more clear for the state mutation, and I wasn't considering known-good. I think the channel for waking up the other goroutines is worth preserving, though. Added more concrete suggestion above.